src/hotspot/share/nmt/nmtTreap.hpp

```diff
@@ -0,0 +1,373 @@
+    } else {
+      // We need
+      //         RIGHT
+      //         |
+      //      LEFT
+      // for the invariant re: priorities to hold.
+      right->_left = merge(left, right->_left DEBUG_ONLY(COMMA recur_count + 1));
+      return right;
+    }
+  }
+
+  static TreapNode* find(TreapNode* node, const K& k DEBUG_ONLY(COMMA int recur_count = 0)) {
+    if (node == nullptr) {
+      return nullptr;
+    }
+
+    int key_cmp_k = COMPARATOR::cmp(node->key(), k);
+
+    if (key_cmp_k == 0) { // key EQ k
+      return node;
+    }
+
+    if (key_cmp_k < 0) { // key LT k
+      return find(node->right(), k DEBUG_ONLY(COMMA recur_count + 1));
+    } else { // key GT k
+      return find(node->left(), k DEBUG_ONLY(COMMA recur_count + 1));
+    }
+  }
+
+#ifdef ASSERT
+  void verify_self() {
+    // A balanced binary search tree should have a depth on the order of log(N).
+    // We take the ceiling of log_2(N + 1) * 2.5 as our maximum bound.
+    // For comparison, a RB-tree has a proven max depth of log_2(N + 1) * 2.
+    const int expected_maximum_depth = ceil((log(this->_node_count+1) / log(2)) * 2.5);
+    // Find the maximum depth through DFS and ensure that the priority invariant holds.
+    int maximum_depth_found = 0;
+
+    struct DFS {
+      int depth;
+      uint64_t parent_prio;
+      TreapNode* n;
+    };
+    GrowableArrayCHeap<DFS, mtNMT> to_visit;
+    constexpr const uint64_t positive_infinity = 0xFFFFFFFFFFFFFFFF;
+
+    to_visit.push({0, positive_infinity, this->_root});
+    while (!to_visit.is_empty()) {
+      DFS head = to_visit.pop();
+      if (head.n == nullptr) continue;
+      maximum_depth_found = MAX2(maximum_depth_found, head.depth);
+
+      assert(head.parent_prio >= head.n->_priority, "broken priority invariant");
+
+      to_visit.push({head.depth + 1, head.n->_priority, head.n->left()});
+      to_visit.push({head.depth + 1, head.n->_priority, head.n->right()});
+    }
+    assert(maximum_depth_found - expected_maximum_depth <= 3,
+           "depth unexpectedly large for treap of node count %d, was: %d, expected between %d and %d",
+           _node_count, maximum_depth_found, expected_maximum_depth - 3, expected_maximum_depth);
+
+    // Visit everything in order, see that the key ordering is monotonically increasing.
+    TreapNode* last_seen = nullptr;
+    bool failed = false;
+    int seen_count = 0;
+    this->visit_in_order([&](TreapNode* node) {
+      seen_count++;
+      if (last_seen == nullptr) {
+        last_seen = node;
+        return;
+      }
+      if (COMPARATOR::cmp(last_seen->key(), node->key()) > 0) {
+        failed = false;
+      }
+      last_seen = node;
+    });
+    assert(seen_count == _node_count, "the number of visited nodes do not match with the number of stored nodes");
+    assert(!failed, "keys was not monotonically strongly increasing when visiting in order");
+  }
+#endif // ASSERT
+
+public:
+  Treap(uint64_t seed = static_cast<uint64_t>(os::random()))
```

Comment:
Do we need 64 bit random number here? If we really need 64 bits shouldn't we do something like:

`  Treap(uint64_t seed = static_cast<uint64_t>(os::random())) | (static_cast<uint64_t>(os::random())) << 32)`

or add `os::random_64bits()` that does exactly this?

Notice that I filed https://bugs.openjdk.org/browse/JDK-8332618 asking to rename `os::random() `to `os::random_32bits()`, so this would go nicely with that change later.

Reply:
I just want to get as much entropy as possible. Yeah, let's do the bit shift inline for now